clients/cobol-lsp-vscode-extension/src/services/copybook/CopybooksCompletionProvider.ts

```
@@ -0,0 +1,114 @@
+import {
+  COPYBOOK_COMPLETIONS_SORT_PREFIX,
+  DEFAULT_DIALECT,
+} from "../../constants";
+import { DialectRegistry } from "../DialectRegistry";
+import { listLocalCopybooks } from "./LocalCopybooksService";
+
+export class CopybooksCompletionProvider implements CompletionItemProvider {
+  constructor(private cds?: CopybookDownloadService) {}
+
+  async provideCompletionItems(
+    document: TextDocument,
+    position: Position,
+    _token: CancellationToken,
+    _context: CompletionContext,
+  ): Promise<CompletionItem[]> {
+    const line = document
+      .lineAt(position.line)
+      .text.slice(0, position.character)
+      .trim();
+
+    const dialectKeywords = DialectRegistry.getActiveDialects()
+      .filter((di) => di.copybooksKeyword)
+      .map((di) => di.copybooksKeyword);
+
+    const COPY_PATTERN = new RegExp(
+      `^.*\\bCOPY(?:\\s+(${dialectKeywords.join("|")}))?(?:\\s+([^\\s]+))?$`,
+      "i",
+    );
+
+    const match = line.match(COPY_PATTERN);
+    if (!match) {
+      return [];
+    }
+
+    const dialectKeyword = match[1];
+    const prefix = match[2]?.replace(`"`, "").replace(`'`, "");
+
+    const dialect =
+      DialectRegistry.identifyDialect(dialectKeyword)?.name ?? DEFAULT_DIALECT;
+
+    let copybooks: string[] = [];
+    let copybooksLoadingFinished = false;
+    const copybooksLoadingPromise = (async () => {
+      copybooks = (
+        await Promise.all([
+          this.cds?.listRemoteCopybooks(document.uri.toString(), dialect) ?? [],
+          listLocalCopybooks(document.uri.toString(), dialect),
+        ])
+      ).flat();
+      copybooksLoadingFinished = true;
+    })();
+    await Promise.race([
+      copybooksLoadingPromise,
+      new Promise((resolve) => {
+        setTimeout(resolve, 200);
+      }),
+    ]);
+
+    // If we are unable to provide completions immediately, show info dialog
+    if (!copybooksLoadingFinished) {
+      await window.withProgress(
```

Let's assume that the request fails - will I be seeing the progress bar any time I type?

---

It depends... if the request fails under 200ms, then no, but if it takes longer, then yes.
Do you think we should show the failed request to the user, so they can fix theirs configuration? Or display this loading dialog just once or not at all and keep it failing silently.

---

We should definitely make sure that we don't cause account lockout if the request is failing due to invalid credentials.

But in general, if the request fails e.g. 5 times in the last minute, disable it and show a dialog that we've done so with an option/command to re-enable it (e.g. once the issue is fixed)?

---

Added check to block zowe requests that keep failing and dialog with error detail and possibility to reenable the similar requests.